ggml/include/ggml.h

```diff
@@ -2009,10 +2024,21 @@ extern "C" {
     GGML_API size_t ggml_graph_overhead(void);
     GGML_API size_t ggml_graph_overhead_custom(size_t size, bool grads);
 
+    GGML_API bool                            ggml_threadpool_params_match (const struct ggml_threadpool_params *p0, const struct ggml_threadpool_params *p1);
+    GGML_API struct ggml_compute_threadpool* ggml_create_threadpool       (struct ggml_threadpool_params  * params);
+    GGML_API void                            ggml_release_threadpool      (struct ggml_compute_threadpool * threadpool);
+    GGML_API int32_t                         ggml_threadpool_get_n_threads(struct ggml_compute_threadpool * threadpool);
+    GGML_API void                            ggml_pause_threadpool        (struct ggml_compute_threadpool * threadpool);
+    GGML_API void                            ggml_resume_threadpool       (struct ggml_compute_threadpool * threadpool);
```

Comment:
It's needed when when more than one threadpool is involved (one for batch and one for non-batch). If we don't start the second pool in paused state there is a regression in batch processing where both pools are running initially (especially if aggressive polling is enabled). I was hunting down the regressions from our earlier tests and some were caused by this overlap.

Reply:
To add to that, user-control for pausing and resuming threadpools is crucial for cases where we have multiple models running back-to-back (e.g. speculative). In cases where the threadpools are running in pure polling mode, we see severe degradation in performance if we have 2 threadpools active at the same time (where one is waiting and the other is actively running ggml_graph_compute)

Reply:
It seems like the simplest way to solve would be to reuse the same threadpool for back-to-back evaluations (including for batch and non-batch, just with a different number of threads). I can understand that with some very complicated hardware with multiple NUMA nodes, it may be desirable to have very fine grained control over which threads are used, but this cannot be done at the expense of making every other use case needlessly more complicated.

Reply:
I agree that no use case should be made more complicated due to threadpooling. Which is why the default behavior now is to let GGML/llama.cpp manage threadpools under the hood. ggml_threadpool design simply allows for fine-grain control, but doesn’t require it.
For example, if you want to run the draft model on cores 0-3 and the target model on cores 2-7, you can now easily do that, and you won’t have to change any other example’s code.

Reply:
It still adds complexity to the user code. In this case llama.cpp is the user code, and other ggml applications will need to do the same.

> For example, if you want to run the draft model on cores 0-3 and the target model on cores 2-7, you can now easily do that

I don't really see why you would want to do that. The main use case that I can see for having different thread pools would be for example running different models simultaneously in a different NUMA node each. However if you think that this is important, I think this logic could still be implemented entirely in the threadpool implementation by automatically pausing other threadpools that use the same threads in `ggml_graph_compute`.

Reply:
NUMA is one use case but there are others. X-Elite has 3x CPU clusters, each cluster has its own L2.
It's not a NUMA system (ie the "distance" to memory is the same) but it does benefit from thread placements.
https://chipsandcheese.com/2024/07/09/qualcomms-oryon-core-a-long-time-in-the-making/

Mobile chipsets (Snapdragon Gen 3, etc) have their own tricks (big beefy core 7, e-cores, etc).

I do agree that somewhat explicit management adds complexity to the user code but at the same time I feel like adding too many smarts to the threadpool itself is less beneficial because an advanced app would know better when and how it wants to switch, and the simple app doesn't really care. 

Here are the use-cases that the current code supports well

1. Basic app : no special requirements
   - This app does not explicitly allocate threadpool
   - We create a temporary (aka disposable) pool inside graph_compute()
   - This is not much different from the original threading or openmp
   - The ggml API is basically the same as before (just need to pass null for threadpool)
2. A bit more advanced app: reduced overhead/latency is required
   - This app allocates a single threadpool with default params
   - The threadpool is reused for all `ggml_graph_compute` (llama_encode/decode, etc)
   - The benefit here is that all of the overhead of starting/stoping threads is gone
   - If the app is optimized for specific platform(s) it might take advantage of the thread placement, polling and priority settings.  
   - The ggml API is very simple, just need to pass n_threads to `ggml_threadpool_params_default()` and allocate / attach the threadpool. 
   - llama-bench is a good example of how to handle this case
3. Advanced app: multiple models, reduced overhead/latency, complex LLM scenarios
   - This app allocates multiple threadpools (batch, non-batch, multiple models)
   - It takes advantage of the finegrain control of the threadpools: advanced CPU placement, threadpool switching, etc
   - `llama-cli` is  a good example of how to handle batch, non-batch usecases and take advantage of the placement, priority, etc.
   - Yes, the `ggml_threadpool` API is more involved here but again building all that into the threadpool and trying to guess what the app really wants seems like an overkill.

Reply:
I would expect case 1 to get the same performance as case 2. As long as the user needs to create a pool manually, this approach will be inherently inferior to OpenMP, which can achieve the same with zero boilerplate added to the application code. Only case 3 should need to manage pools manually, and even then, that's not an excuse to pool API more complicated than it strictly needs to be. I would be much happier about this if the changes to `llama.cpp` to deal with switching pools, or the changes to `llama-bench` to reset pools after every test, could be removed.

Reply:
I'm working on moving threadpool switching into ggml. 
`src/llama.cpp` just needs to set which threadpool to use for next `graph_compute`, same as it does with setting `n_threads` vs `n_threads_batch`. 

For `llama-bench` I mentioned it way back in this PR while hunting down regressions. We used to just start a single threadpool and reuse between the tests, but that's suboptimial, also it doesn't allow for completely different cpumask  between the tests.
For instance something like this

$ ./build-arm64-windows-llvm-release/bin/llama-bench.exe -m ../gguf/llama-v3.1.q4_0_4_8.gguf -t 4 --cpu-mask 0xf00,0xf0,0xf,0x311 --cpu-strict 1 -p 64 -n 16 --prio 1 --delay 30
| model                          |       size |     params | backend    | threads | cpu_mask   | cpu_strict |          test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ------: | ---------- | ---------: | ------------: | -------------------: |
| llama 8B Q4_0_4_8              |   4.33 GiB |     8.03 B | CPU        |       4 | 0xf00      |          1 |          pp64 |         85.72 ± 0.42 |
| llama 8B Q4_0_4_8              |   4.33 GiB |     8.03 B | CPU        |       4 | 0xf00      |          1 |          tg16 |         17.05 ± 0.02 |
| llama 8B Q4_0_4_8              |   4.33 GiB |     8.03 B | CPU        |       4 | 0xf0       |          1 |          pp64 |         85.64 ± 0.38 |
| llama 8B Q4_0_4_8              |   4.33 GiB |     8.03 B | CPU        |       4 | 0xf0       |          1 |          tg16 |         16.97 ± 0.02 |
| llama 8B Q4_0_4_8              |   4.33 GiB |     8.03 B | CPU        |       4 | 0xf        |          1 |          pp64 |         79.05 ± 0.81 |
| llama 8B Q4_0_4_8              |   4.33 GiB |     8.03 B | CPU        |       4 | 0xf        |          1 |          tg16 |         14.30 ± 0.02 |
| llama 8B Q4_0_4_8              |   4.33 GiB |     8.03 B | CPU        |       4 | 0x311      |          1 |          pp64 |         85.08 ± 0.20 |
| llama 8B Q4_0_4_8              |   4.33 GiB |     8.03 B | CPU        |       4 | 0x311      |          1 |          tg16 |         21.03 ± 0.00 |

build: 3bcc4dee (3674)

Rebinding the threads automatically, etc is going to add a bunch of logic and corner cases.
Note that OpenMP can't do any of that. I don't think allowing explicit control for the app is inferior to OpenMP, quite the opposite. 

Anyway, take a look at the next updates to move the switching logic.
Hopefully we can merge that and then iterate further on moving threading stuff into `ggml-threads.cpp/h`, making it more generic and adding support for reusing temporary threadpools (ie zero-boiler plate case).